console/v1/types_console_sample.go

```diff
@@ -0,0 +1,166 @@
+	// Most icons will be quadratic but the web console ensures that all aspect ratios work.
+	// +kubebuilder:validation:Optional
+	// +optional
+	Icon string `json:"icon,omitempty"`
+
+	// provider like "Red Hat" is optional and shown with the prefix "Provided by ".
+	//
+	// Currently, is only shown on the sample card tile below the title.
+	//
+	// It is optional and must be no more than 20 characters in length.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:MaxLength=20
+	// +optional
+	Provider string `json:"provider"`
+
+	// badge is a short visual annotation to group samples.
+	//
+	// It is optional and must be no more than 20 characters in length.
+	//
+	// Recommendation is a singular term like "Builder Image", "Devfile" or "Serverless Function".
+	//
+	// Currently, the badge is displayed on sample card tile in the top right corner.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:MaxLength=20
+	// +optional
+	Badge string `json:"badge"`
+
+	// tags like "Java", "Quarkus", etc.
+	// Tags are optional and can be used by the user to find samples in the samples catalog.
+	// They will be also displayed on the samples details page.
+	// +kubebuilder:validation:Optional
+	// +optional
+	Tags []string `json:"tags"`
+
+	// The actual sample source, like an external git repository or container image.
+	// +kubebuilder:validation:Required
+	// +required
+	Source ConsoleSampleSource `json:"source"`
+}
+
+// ConsoleSampleSourceType is an enumeration of the supported sample types.
+// Unsupported samples types will be ignored in the web console.
+// +kubebuilder:validation:Enum:=GitImport;ContainerImport
+type ConsoleSampleSourceType string
+
+const (
+	// A sample that let the user import code from a git repository.
+	GitImport ConsoleSampleSourceType = "GitImport"
+	// A sample that let the user import a container image.
+	ContainerImport ConsoleSampleSourceType = "ContainerImport"
+)
+
+// ConsoleSampleSource is the actual sample definition and can hold different sample types.
+// Unsupported sample types will be ignored in the web console.
+// +union
+// +kubebuilder:validation:XValidation:rule="self.type == 'GitImport' ? has(self.gitImport) : !has(self.gitImport)",message="source.gitImport is required when source.type is GitImport, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="self.type == 'ContainerImport' ? has(self.containerImport) : !has(self.containerImport)",message="source.containerImport is required when source.type is ContainerImport, and forbidden otherwise"
+type ConsoleSampleSource struct {
+	// type of the sample, currently supported: "GitImport";"ContainerImport"
+	// +unionDiscriminator
+	// +kubebuilder:validation:Enum:="GitImport";"ContainerImport"
+	// +kubebuilder:validation:Required
+	Type ConsoleSampleSourceType `json:"type"`
+
+	// gitimport let the user import code from a git repository.
+	// +unionMember,optional
```

Comment:
This has a different semantic to, please swap
```suggestion
	// +unionMember
	// +optional
```